File: services/engine/src/engine/api/schemas/market.py
from datetime import datetime

from pydantic import BaseModel


class BarData(BaseModel):
    symbol: str
    market: str
    timeframe: str
    open: float
    high: float
    low: float
    close: float
    volume: float
    timestamp: datetime


class MarketDataQuery(BaseModel):
    symbol: str
    market: str = "crypto"
    timeframe: str = "1d"
    start: datetime | None = None
    end: datetime | None = None
    limit: int = 100


class LatestQuote(BaseModel):
    symbol: str
    market: str
    price: float
    change_24h: float
    volume_24h: float
    timestamp: datetime


class SymbolInfo(BaseModel):
    symbol: str
    market: str
    base: str
    quote: str
    status: str = "active"


class MarketDataResponse(BaseModel):
    symbol: str
    timeframe: str
    bars: list[BarData] = []

File: apps/web/src/types/market.ts
export const Market = {
  A_STOCK: 'a_stock',
  CRYPTO: 'crypto',
  US_STOCK: 'us_stock',
} as const
export type Market = (typeof Market)[keyof typeof Market]

export const TimeFrame = {
  TICK: 'tick',
  M1: '1m',
  M5: '5m',
  M15: '15m',
  H1: '1h',
  H4: '4h',
  D1: '1d',
  W1: '1w',
} as const
export type TimeFrame = (typeof TimeFrame)[keyof typeof TimeFrame]

export interface Bar {
  symbol: string
  market: Market
  timeframe: TimeFrame
  open: number
  high: number
  low: number
  close: number
  volume: number
  timestamp: string
}

export interface Tick {
  symbol: string
  market: Market
  price: number
  volume: number
  timestamp: string
}
